### PR TITLE
Update mafft

### PR DESCRIPTION
**Pull request by DamCorreia**
Hi,
Here is a proposition to add new features on mafft.xml

+ add user custom parameters
+ add more options and associated help
+ change method to modify output format

What do you think about this changes ?

### CODE REVIEW

File: tools/mafft/mafft.xml [hunk 1]
```diff
@@ -65,36 +82,76 @@
...
+            </when>
         </conditional>
+        <param name="op" type="float" value="1.53" label="Gap opening penalty" help="1.53 default value" />
         <param name="adjustdirection" type="select" display="radio" label="Direction of nucleotide sequences" help="Generate reverse complement sequences, as necessary, and align them together with the remaining sequences">
             <option value="--adjustdirection">adjust direction</option>
             <option value=" " selected="true">do not adjust direction</option>
-            </param>
+        </param>
+        <conditional name="matrix_condition">
+            <param name="use_matrix" type="boolean" truevalue="true" falsevalue="false" checked="False" label="Use matrix ?" help="Usefull only for amino acids" />
```
Comment (bgruening) [inline tools/mafft/mafft.xml]: Can we change this to a select box, please. 
This is recommended by the current best practice guide: http://galaxy-iuc-standards.readthedocs.org/en/latest/best_practices/tool_xml.html?highlight=conditionals#booleans

File: tools/mafft/mafft.xml [hunk 2]
```diff
@@ -65,36 +82,76 @@
...
+            <option value="--clustalout">ClustalW</option>
+            <option value="--phylipout">Phylip</option>
         </param>
     </inputs>
     <outputs>
-        <data format="fasta" name="outputFasta" label="${tool.name} on ${on_string}">
-            <filter>outputFormat == 'fasta'</filter>
+        <data format="fasta" name="outputAlignment" label="${tool.name} on ${on_string}">
+            <change_format>
+                <when input="outputFormat" value="--clustalout" format="clustal"/>
```
Comment (bgruening) [inline tools/mafft/mafft.xml]: Why does `--clustalout` output a FASTA file?

**Comment by bgruening** [comment]
@DamCorreia This looks great! Test data hasn't changed?
Thanks very much for this contribution!

xref: https://github.com/bgruening/galaxytools/issues/290

File: tools/mafft/mafft.xml [hunk 3]
```diff
@@ -65,36 +82,76 @@
             <when value="mafft-linsi"/>
             <when value="mafft-qinsi"/>
             <when value="mafft-xinsi"/>
+            <when value="custom">
+                <param name="distance_method" type="select" display="radio" label="Distance method" help="Distance method must be chosen regarding your data">
+                    <option value="6merpair" selected="true">Shared 6mers distance (fastest)</option>
+                    <option value="globalpair">Global alignment (NW)</option>
+                    <option value="localpair">Local alignment (SW)</option>
```
Comment (nsoranzo) [inline tools/mafft/mafft.xml]: I'd be explicit here: NW -> Needleman-Wunsch, SW -> Smith-Waterman

File: tools/mafft/mafft.xml [hunk 4]
```diff
@@ -16,35 +16,52 @@
     <![CDATA[
         #if $cond_flavour.flavourType != 'custom'
             $cond_flavour.flavourType
+            
         #elif $cond_flavour.flavourType == 'custom'
             ### full parameter options
             mafft
+            --$cond_flavour.distance_method
```
Comment (nsoranzo) [inline tools/mafft/mafft.xml]: I'd move the `--` to the `<option>` values for consistency.

**Comment by nsoranzo** [comment]
I don't know what happened, but there seems to be conflicts which prevent the merge of the PR.

File: tools/mafft/mafft.xml [hunk 5]
```diff
@@ -65,36 +80,73 @@
...
             <when value="mafft-qinsi"/>
             <when value="mafft-xinsi"/>
+            <when value="custom">
+                <param name="distance_method" type="select" display="radio" label="Distance method" help="Distance method must be chosen regarding your data">
+                    <option value="--6merpair" selected="true">Shared 6mers distance (fastest)</option>
+                    <option value="--globalpair">Global alignment (Needleman-Wunsch)</option>
+                    <option value="--localpair">Local alignment (Smith-Waterman)</option>
+                    <option value="--genafpair">Local, affine gap cost</option>
+                </param>
+                <param name="ep" type="float" value="0.123" label="Gap extend penalty" help="0.123 default value" />
```
Comment (nsoranzo) [inline tools/mafft/mafft.xml]: Regarding this label, the man page describe "ep" as "Offset value, which works like gap extension penalty, for group-to-group alignment", which seems more accurate. Also you can remove the default value from the `help` attribute.

File: tools/mafft/mafft.xml [hunk 6]
```diff
@@ -65,36 +80,73 @@
...
+                <param name="retree" type="integer" value="2" min="1" max="100" label="Guide tree is built number times in the progressive stage." help="Valid with 6mer distance" />
```
Comment (nsoranzo) [inline tools/mafft/mafft.xml]: s/built number times/built this number of times/

File: tools/mafft/mafft.xml [hunk 7]
```diff
@@ -65,36 +80,73 @@
...
+                <param name="iterations" type="integer" value="1" min="0" max="1000" label="Maximum number of iterations" help="1000 for maximum quality" />
```
Comment (nsoranzo) [inline tools/mafft/mafft.xml]: Default is 0 according to `mafft --help` and the man page.

File: tools/mafft/mafft.xml [hunk 8]
```diff
@@ -16,35 +16,50 @@
...
+        $reorder
+        $getTree    
+        $outputFormat.value
+        $inputSequences > $outputAlignment;
         
-        #if $outputFormat.value == 'fasta'
-            $outputFasta
-        #elif $outputFormat.value == 'clustalw'
-            $outputClustalW
+        #if $getTree.value == True:
```
Comment (nsoranzo) [inline tools/mafft/mafft.xml]: I'd use:

`#if $getTree`

File: tools/mafft/mafft.xml [hunk 9]
```diff
@@ -16,35 +16,50 @@
...
+        #if $matrix_condition.matrix == "BLOSUM"
+          --bl ${matrix_condition.BLOSUM}
+        #elif $matrix_condition.matrix == "PAM"
+           --jtt ${matrix_condition.PAM}
         #end if
         
-        $inputSequences > 
+        $reorder
+        $getTree    
+        $outputFormat.value
```
Comment (nsoranzo) [inline tools/mafft/mafft.xml]: `$outputFormat`

File: tools/mafft/mafft.xml [hunk 10]
```diff
@@ -1,4 +1,4 @@
-<tool id="rbc_mafft" name="MAFFT" version="7.221.1">
+<tool id="rbc_mafft" name="MAFFT" version="7.221.2">
     <description>Multiple alignment program for amino acid or nucleotide sequences</description>
     <requirements>
         <requirement type="package" version="7.221">mafft</requirement>
```
Comment (nsoranzo) [inline tools/mafft/mafft.xml]: Maybe a good time to require the latest 7.271 version? In this case, tool_dependencies.xml should be updated accordingly.

**Comment by nsoranzo** [comment]
Looks good to me, but there seems to still be some conflict which prevents the merge, maybe try to rebase?

@hrhotz may always want to review this.

**Comment by hrhotz** [comment]
phew, too many changes for me to do a quick review....looking at "6b89e08", I am alreday struggling with the help: there are now much more options listed, than actually offered, even if 'custom' is selected, aren't-they?   sorry, I can't be of more help today

**Comment by DamCorreia** [comment]
@bgruening Tests and test-data haven't changed and they still work

**Comment by DamCorreia** [comment]
@nsoranzo Thank you for your advises, I made a rebase and I have solved conflicts

File: tools/mafft/mafft.xml [hunk 11]
```diff
@@ -60,35 +80,73 @@
...
+                    <option value="--globalpair">Global alignment (Needleman-Wunsch)</option>
+                    <option value="--localpair">Local alignment (Smith-Waterman)</option>
+                    <option value="--genafpair">Local, affine gap cost</option>
+                </param>
+                <param name="ep" type="float" value="0.123" label="Offset value" help="Offset value, which works like gap extension penalty, for group-to-group alignment" />
+                <param name="retree" type="integer" value="2" min="1" max="100" label="Guide tree is built this number of times in the progressive stage." help="Valid with 6mer distance" />
+                <param name="iterations" type="integer" value="0" min="0" max="1000" label="Maximum number of iterations" help="1000 for maximum quality" />
+            </when>
         </conditional>
+        <param name="op" type="float" value="1.53" label="Gap opening penalty" help="1.53 default value" />
```
Comment (nsoranzo) [inline tools/mafft/mafft.xml]: Is there a reason for "ep" being inside custom MAFFT flavour while "op" is here?

**Comment by bgruening** [comment]
@DamCorreia this looks good to me, can you please increase the version number and we are ready :)

**Comment by DamCorreia** [comment]
@bgruening It's done :)

**Comment by bgruening** [comment]
Thanks!

**Comment by bgruening** [comment]
Updated the TSs!

**Pull request merged**